kong/router/transform.lua

```diff
@@ -0,0 +1,638 @@
+
+
+local is_null
+local is_empty_field
+do
+  local null    = ngx.null
+  local isempty = require("table.isempty")
+
+  is_null = function(v)
+    return v == nil or v == null
+  end
+
+  is_empty_field = function(f)
+    return f == nil or f == null or isempty(f)
+  end
+end
+
+
+local function escape_str(str)
+  -- raw string
+  if not str:find([["#]], 1, true) then
+    return "r#\"" .. str .. "\"#"
+  end
+
+  -- standard string escaping (unlikely case)
+  if str:find([[\]], 1, true) then
+    str = str:gsub([[\]], [[\\]])
+  end
+
+  if str:find([["]], 1, true) then
+    str = str:gsub([["]], [[\"]])
+  end
+
+  return "\"" .. str .. "\""
+end
+
+
+-- split port in host, ignore form '[...]'
+-- example.com:123 => example.com, 123
+-- example.*:123 => example.*, 123
+local split_host_port
+do
+  local tonumber = tonumber
+
+  local DEFAULT_HOSTS_LRUCACHE_SIZE = utils.DEFAULT_MATCH_LRUCACHE_SIZE
+
+  local memo_hp = lrucache.new(DEFAULT_HOSTS_LRUCACHE_SIZE)
+
+  split_host_port = function(key)
+    if not key then
+      return nil, nil
+    end
+
+    local m = memo_hp:get(key)
+
+    if m then
+      return m[1], m[2]
+    end
+
+    local p = key:find(":", nil, true)
+    if not p then
+      memo_hp:set(key, { key, nil })
+      return key, nil
+    end
+
+    local port = tonumber(key:sub(p + 1))
+
+    if not port then
+      memo_hp:set(key, { key, nil })
+      return key, nil
+    end
+
+    local host = key:sub(1, p - 1)
+
+    memo_hp:set(key, { host, port })
+
+    return host, port
+  end
+end
+
+
+local LOGICAL_OR  = " || "
+local LOGICAL_AND = " && "
+
+
+local OP_EQUAL    = "=="
+local OP_PREFIX   = "^="
+local OP_POSTFIX  = "=^"
+local OP_REGEX    = "~"
+local OP_IN       = "in"
+
+
+local DOT              = byte(".")
+local TILDE            = byte("~")
+local ASTERISK         = byte("*")
+
+
+-- reuse buffer objects
+local values_buf        = buffer.new(64)
+local nets_buf          = buffer.new(64)
+local expr_buf          = buffer.new(64)
+local hosts_buf         = buffer.new(64)
+local headers_buf       = buffer.new(64)
+local single_header_buf = buffer.new(64)
+
+
+-- sep: a seperator of expressions, like '&&'
+-- idx: indicate whether or not to add 'sep'
+--      for example, we should not add 'sep' for the first element in array
+local function expression_append(buf, sep, str, idx)
+  if #buf > 0 and
+     (idx == nil or idx > 1)
+  then
+    buf:put(sep)
+  end
+  buf:put(str)
+end
+
+
+local function gen_for_field(name, op, vals, val_transform)
+  if is_empty_field(vals) then
+    return nil
+  end
+
+  local vals_n = #vals
+  assert(vals_n > 0)
+
+  values_buf:reset():put("(")
+
+  for i = 1, vals_n do
+    local p = vals[i]
+    local op = (type(op) == "string") and op or op(p)
+
+    local expr = fmt("%s %s %s", name, op,
+                    escape_str(val_transform and val_transform(op, p) or p))
+
+    expression_append(values_buf, LOGICAL_OR, expr, i)
```

Comment:
comments for reviewers, Line 152~155 changed from 

```
 if i > 1 then
      values_buf:put(LOGICAL_OR)
    end

    values_buf:putf("%s %s %s", name, op,
                    escape_str(val_transform and val_transform(op, p) or p))
```

Reply:
Yes, it is worthy to be noticed.